pkg/services/oauthserver/oasimpl/service.go

```diff
@@ -0,0 +1,504 @@
+	// client_secret must be at least 32 bytes long
+	secret, err := s.randString(32)
+	if err != nil {
+		return "", "", err
+	}
+	return id, secret, err
+}
+
+func (s *OAuth2ServiceImpl) computeGrantTypes(selfAccessEnabled, impersonationEnabled bool) []string {
+	grantTypes := []string{}
+
+	// If the client has permissions, it can use the client credentials grant type
+	if selfAccessEnabled {
+		grantTypes = append(grantTypes, string(fosite.GrantTypeClientCredentials))
+	}
+
+	// If the client has impersonate permissions, it can use the JWT bearer grant type
+	if impersonationEnabled {
+		grantTypes = append(grantTypes, string(fosite.GrantTypeJWTBearer))
+	}
+
+	return grantTypes
+}
+
+func (s *OAuth2ServiceImpl) handleKeyOptions(ctx context.Context, keyOption *oauthserver.KeyOption) (*oauthserver.KeyResult, error) {
+	if keyOption == nil {
+		return nil, fmt.Errorf("keyOption is nil")
+	}
+
+	var publicPem, privatePem string
+
+	if keyOption.Generate {
+		switch s.cfg.OAuth2ServerGeneratedKeyTypeForClient {
+		case "RSA":
+			privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+			if err != nil {
+				return nil, err
+			}
+			publicPem = string(pem.EncodeToMemory(&pem.Block{
+				Type:  "RSA PUBLIC KEY",
+				Bytes: x509.MarshalPKCS1PublicKey(&privateKey.PublicKey),
+			}))
+			privatePem = string(pem.EncodeToMemory(&pem.Block{
+				Type:  "RSA PRIVATE KEY",
+				Bytes: x509.MarshalPKCS1PrivateKey(privateKey),
+			}))
+			s.logger.Debug("RSA key has been generated")
+		default: // default to ECDSA
+			privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+			if err != nil {
+				return nil, err
+			}
+			publicDer, err := x509.MarshalPKIXPublicKey(&privateKey.PublicKey)
+			if err != nil {
+				return nil, err
+			}
+
+			privateDer, err := x509.MarshalPKCS8PrivateKey(privateKey)
+			if err != nil {
+				return nil, err
+			}
+
+			publicPem = string(pem.EncodeToMemory(&pem.Block{
+				Type:  "PUBLIC KEY",
+				Bytes: publicDer,
+			}))
+			privatePem = string(pem.EncodeToMemory(&pem.Block{
+				Type:  "PRIVATE KEY",
+				Bytes: privateDer,
+			}))
+			s.logger.Debug("ECDSA key has been generated")
+		}
+
+		return &oauthserver.KeyResult{
+			PrivatePem: privatePem,
+			PublicPem:  publicPem,
+			Generated:  true,
+		}, nil
+	}
+
+	// TODO MVP allow specifying a URL to get the public key
+	// if registration.Key.URL != "" {
+	// 	return &oauthserver.KeyResult{
+	// 		URL: registration.Key.URL,
+	// 	}, nil
+	// }
+
+	if keyOption.PublicPEM != "" {
+		pemEncoded, err := base64.StdEncoding.DecodeString(keyOption.PublicPEM)
+		if err != nil {
+			s.logger.Error("cannot decode base64 encoded PEM string", "error", err)
+		}
+		_, err = utils.ParsePublicKeyPem(pemEncoded)
+		if err != nil {
+			s.logger.Error("cannot parse PEM encoded string", "error", err)
+			return nil, err
+		}
+		return &oauthserver.KeyResult{
+			PublicPem: string(pemEncoded),
+		}, nil
+	}
+
+	return nil, fmt.Errorf("at least one key option must be specified")
```

Comment:
Same here: I think it's fine, if impersonation is not needed, not to generate the key.